File: week2/vigenere.c
#include <cs50.h>
#include <stdio.h>
#include <ctype.h>
#include <string.h>

int shift(char c);
char encode(char letter, int lock);

int main(int argc, string argv[])
{
    // check only 2 arguments
    if (argc == 2)
    {
        //iterate through 2nd argument and check all chars are digits
        for (int i = 0; i < strlen(argv[1]); i++)
        {
            if (isalpha(argv[1][i]) == 0)
            {
                printf("Usage: ./vigenere keyword\n");
                return 1;
            }
        }
        string plain = get_string("plaintext: ");
        printf("ciphertext: ");
        // set amount of 
        int shiftBy = -1;
        for (int i = 0; i < strlen(plain); i++)
        {
            // reset keyIndex tracker to 0
            int keyIndex = 0;
            if (shiftBy < (strlen(argv[1]) - 1))
            {
                // increase shiftBy by 1 and add to keyIndex tracker
                // if shiftBy < n(strlenargv[1]) - 1  
                shiftBy++;                
                keyIndex += shiftBy;
            }
            else 
            {
                // reset counters to 0 when counter reaches argument length
                keyIndex = 0;
                shiftBy = 0;
            }
            int key = shift(argv[1][keyIndex]);
            // check character is a letter and encode if it is
            if ((plain[i] >= 'a' && plain[i] <= 'z') || (plain[i] >= 'A' && plain[i] <= 'Z'))
            {
                printf("%c", encode(plain[i], key)); // print character using encode function
            }
            else 
            {
                // print character unciphered if it isn't a letter
                printf("%c", plain[i]);
                shiftBy--;
                keyIndex -= shiftBy;
            }
        }
        printf("\n");
        return 0;
    }
    else 
    {
        printf("Usage: ./vigenere keyword\n");
        return 1;
    }
}

int shift(char c)
{
    // check if upper or lowercase - A is less than a
    if (c >= 'a')
    {
        int shifted = c - 'a';
        return shifted;
    }
    else
    {
        int shifted = c - 'A';
        return shifted;
    }
}

char encode(char letter, int lock)
{
    // check if upper or lowercase - A is less than a
    if (letter >= 'a')
    {
        // subtract value of a from letter and add key
        // if value is > 26, remainder causes cipher to 'loop back' to a
        // add 0 - 26 value to value of a to get ciphered char
        char ciphered = (((letter - 'a') + lock) % 26) + 'a';
        return ciphered; 
    }
    else 
    {
        char ciphered = (((letter - 'A') + lock) % 26) + 'A';
        return ciphered;
    }
}

File: week6/vigenere/vigenere.py
import sys
import cs50

# check for only 2 arguments and exit if not
if len(sys.argv) is not 2:
    print("Usage: python vigenere.py k")
    sys.exit(1)

# check key is only made up of alphabetical characters and exit if not
if sys.argv[1].isalpha() == False:
    print("Usage: python vigenere.py k")
    sys.exit(1)

# convert character to positional integer value a/A = 0 b/B = 1 etc


def shift(c):
    if c.isupper() == True:
        c = ord(c) - 65
    else:
        c = ord(c) - 97
    return c


# store argument keys
keys = sys.argv[1]

keyList = []

# store each key in keyList
for i in keys:
    keyList.append(shift(i))

keyLength = len(keyList)

# get plaintext
plain = cs50.get_string("plaintext: ")

# define encode function


def encode(l, k):
    # convert letter to int from 0-26 store in val & return
    if l.isupper() == True:
        val = (((ord(l) - 65) + k) % 26) + 65
    else:
        val = (((ord(l) - 97) + k) % 26) + 97

    return chr(val)


# empty list for ciphered letters
ciphered = []
# set counter for keyList index
curr = -1

for i in plain:
    # + 1 to keyIndex for each letter of plain & return to 0th index once length of keyList reached
    keyIndex = 0
    if curr < len(keyList) - 1:
        curr += 1
        keyIndex += curr
    else:
        curr = 0
        keyIndex = 0
    # set key variable to current keyIndex
    key = keyList[keyIndex]
    # use encode function to cipher current letter by keyIndex
    c = encode(i, key)
    if i.isalpha() == True:
        ciphered.append(c)
    # if current char is not a letter do not cipher and reduce key index & counter by 1
    else:
        ciphered.append(i)
        curr -= 1
        keyIndex -= curr

# join ciphered list of letters into a single string & print
res = "".join(ciphered)
print(f"ciphertext: {res}")